src/Symfony/Component/HttpKernel/Tests/EventListener/SessionListenerTest.php

```diff
@@ -33,6 +34,101 @@
+     */
+    public function testSessionCookieOptions(array $phpSessionOptions, array $sessionOptions, array $expectedSessionOptions)
+    {
+        $session = $this->getMockBuilder(Session::class)->disableOriginalConstructor()->getMock();
+        $session->expects($this->exactly(2))->method('getUsageIndex')->will($this->onConsecutiveCalls(0, 1));
+        $session->expects($this->exactly(1))->method('getId')->willReturn('123456');
+        $session->expects($this->exactly(1))->method('getName')->willReturn('PHPSESSID');
+        $session->expects($this->exactly(1))->method('save');
+        $session->expects($this->exactly(1))->method('isStarted')->willReturn(true);
+
+        if (isset($phpSessionOptions['samesite'])) {
+            ini_set('session.cookie_samesite', $phpSessionOptions['samesite']);
+        }
+        session_set_cookie_params(0, $phpSessionOptions['path'] ?? null, $phpSessionOptions['domain'] ?? null, $phpSessionOptions['secure'] ?? null, $phpSessionOptions['httponly'] ?? null);
+
+        $container = new Container();
+        $container->set('initialized_session', $session);
+
+        $listener = new SessionListener($container, false, $sessionOptions);
+        $kernel = $this->getMockBuilder(HttpKernelInterface::class)->disableOriginalConstructor()->getMock();
+
+        $request = new Request();
+        $listener->onKernelRequest(new RequestEvent($kernel, $request, HttpKernelInterface::MAIN_REQUEST));
+
+        $response = new Response();
+        $listener->onKernelResponse(new ResponseEvent($kernel, new Request(), HttpKernelInterface::MAIN_REQUEST, $response));
+
+        $cookies = $response->headers->getCookies();
+        $this->assertSame('PHPSESSID', $cookies[0]->getName());
+        $this->assertSame('123456', $cookies[0]->getValue());
+        $this->assertSame($expectedSessionOptions['cookie_path'], $cookies[0]->getPath());
+        $this->assertSame($expectedSessionOptions['cookie_domain'], $cookies[0]->getDomain());
+        $this->assertSame($expectedSessionOptions['cookie_secure'], $cookies[0]->isSecure());
+        $this->assertSame($expectedSessionOptions['cookie_httponly'], $cookies[0]->isHttpOnly());
+        $this->assertSame($expectedSessionOptions['cookie_samesite'], $cookies[0]->getSameSite());
+    }
+
+    public function provideSessionOptions(): array
+    {
+        $provideArr = [];
+
+        if (\PHP_VERSION_ID > 70300) {
+            $provideArr = [
+                'set_samesite_by_php' => [
+                    'phpSessionOptions' => ['samesite' => Cookie::SAMESITE_STRICT],
+                    'sessionOptions' => ['cookie_path' => '/test/', 'cookie_domain' => '', 'cookie_secure' => true, 'cookie_httponly' => true],
+                    'expectedSessionOptions' => ['cookie_path' => '/test/', 'cookie_domain' => '', 'cookie_secure' => true, 'cookie_httponly' => true, 'cookie_samesite' => Cookie::SAMESITE_STRICT],
+                ],
+            ];
+        }
```

Comment:
This `$provideArr` structure is unnecessarily complicated. Can we work with a generator instead?

```suggestion
    public function provideSessionOptions(): \Generator
    {
        if (\PHP_VERSION_ID > 70300) {
            yield 'set_samesite_by_php' => [
                'phpSessionOptions' => ['samesite' => Cookie::SAMESITE_STRICT],
                'sessionOptions' => ['cookie_path' => '/test/', 'cookie_domain' => '', 'cookie_secure' => true, 'cookie_httponly' => true],
                'expectedSessionOptions' => ['cookie_path' => '/test/', 'cookie_domain' => '', 'cookie_secure' => true, 'cookie_httponly' => true, 'cookie_samesite' => Cookie::SAMESITE_STRICT],
            ];
        }
```

Reply:
You need to convert the other cases as well. The diff was too large for a change suggestion. 😬